### ISSUE TITLE
Request for Comment: Intro to Programming Courses

### ISSUE THREAD

**Issue by waciumawanjohi** [ossu/computer-science#589]
**Problem:**
The first courses in the curriculum are not appropriate for students who do not know how to program.

**Duration:**
September 30, 2019

**Current State:**
Awaiting feedback

**Proposer:**
Waciuma Wanjohi

**Details:**
OSSU lists prerequisites for students. These are all high school level math and science concepts. However, the first course of the curriculum demands a strong grasp of programming. As stated by @bradleygrant, "[The first course's] biggest downside is it assumes that the student knows Python inside and out by Week 2, which is pretty much unattainable for a first-timer -- especially since the class does little to get you there."

OSSU promises to detail a "Path to a free self-taught education in Computer Science!" That path should be clear about what preparation students need to undertake before starting on the path. Either OSSU must put in clear prerequisites for knowing how to program or OSSU must provide a clear path for beginners to learn how to program.

**Proposal:**
Give students a choice between three different courses that focused on teaching programming.
- [Computing in Python I: Fundamentals and Procedural Programming](https://www.edx.org/course/computing-in-python-i-fundamentals-and-procedural-programming-2)
- [An Introduction to Interactive Programming in Python](https://www.coursera.org/learn/interactive-python-1)
- [Programming for Everybody](https://www.coursera.org/learn/python)

Each of these courses is an introductory programming course from a strong University. None assume prior experience coding. Together these courses have received over 10,000 online reviews on an independent platform.  [Each has an average rating above 4.5 stars](https://www.classcentral.com/report/top-moocs-2019-edition/).

Commenters may wonder at the decision to recommend a choice between three courses rather than a single course. The primary reason is that the introductory courses tend to generate the most debate among OSSU students. De-emphasizing any particular choice should make clear that students are welcome to choose whichever course they think will best suit. Ultimately, a student's success will be determined not by which course they chose first, but by their perseverance through the dozens of courses that follow.

**Alternatives:**
OSSU can list basic understanding of programming as a prerequisite. This would allow the first course to be an introduction to computer science rather than an introduction to programming.

**Comment by Zaris47** [comment]
Harvard's CS50 introduction to computer programming course is pretty good actually. I used that as my first CS course and its very beginners friendly and comprehensive. It's there on Edx.

**Comment by cipherblock** [comment]
> Harvard's CS50 introduction to computer programming course is pretty good actually. I used that as my first CS course and its very beginners friendly and comprehensive. It's there on Edx.

Am I mistaken or is this part of the curriculum under 'Core CS: Core Systems'?

**Comment by Zaris47** [comment]
> > Harvard's CS50 introduction to computer programming course is pretty good actually. I used that as my first CS course and its very beginners friendly and comprehensive. It's there on Edx.
> 
> Am I mistaken or is this part of the curriculum under 'Core CS: Core Systems'?

Oh my mistake, sorry. I just went to look and indeed it is a part of the curriculum. I don't know why we need two introductory courses though.

**Comment by cipherblock** [comment]
I am just starting out on my CS journey and found the OSSU not long ago. I enrolled in 'Introduction to Computer Science and Programming Using Python' and the course is due to start in two days time (28th August). Is it my understanding then that this would not be suitable for me as I have no prior knowledge?

**Comment by bradleygrant** [comment]
I was the one who originally wrote the referenced comment:

"[The first course's] biggest downside is it assumes that the student knows
Python inside and out by Week 2, which is pretty much unattainable for a
first-timer -- especially since the class does little to get you there."

That was based on my experience taking the (MIT Intro To Computing with
Python) course last year, after having already had the experience taking
the Georgia Tech CS1301 course series. It was easy for me, but I'd had the
benefit of having taken a pre-course in Python syntax and control flow.

You have time, assuming you start right now and can devote the next 2 weeks
to getting through Parts 1 and 2 of the CS1301 course (and completing Part
3 concurrently with the MIT course). (All you're doing is completing a
guided study of basically what MIT would have you do anyway...they
seriously want you up to speed on basic Python in about 2 weeks.)

Still, the entire point of THIS PR is to prevent others like you from
having to scramble in order to prepare to take what is ostensibly the
introductory course into the curriculum...

**Comment by cipherblock** [comment]
@bradleygrant Thank you for your response. Is the CS1301 course you're recommending the one on edX? I'm not sure if it was a free course when you took it but I believe it costs £291 now. Would you have any other suggestions for a course I should look at before attempting the MIT course (ideally something free or less expensive than the Georgia Tech one)?

**Comment by KathrynN** [comment]
Hi @bradleygrant 
Have you taken an edX course lately?
They threaten that your access to a course will end on X date - unless you pay in two weeks for the certificate version, of course.  In the courses I've looked at, you tend to get 2 months of a course before it is closed to you forever.
From your comment, this does not match your experience?
Kat

**Comment by cipherblock** [comment]
@bradleygrant I really appreciate you taking the time out to check all of that for me and answer my questions. As you've mentioned the time limits, the MITx course is due to begin tomorrow. So with that in mind do you think I'd still be ok, time wise, to take parts 1 and 2 of the GT course, then come back and begin the MITx course while concurrently finishing up parts 3 and 4 of the GT course?

**Comment by jhubler99** [comment]
Hi @benbrandt 

It seems every course on edX has been paywalled on graded assignments. For example: I completed the course "Introduction to Python: Absolute Beginner" with full access to the graded assignments for free in 2018. Just to see if it is still free, I enrolled in it now and the graded assignments are paywalled. It seems every course on edX are going this route now. This course used electronic grading, so you can get the mark immediately. I found the graded assignments motivated me to complete the course. I think everyone likes to know what their doing wrong. 

This picture is in "How to Code: Simple Data"

![How to Code ](https://user-images.githubusercontent.com/40548778/63901669-c8ba4600-c9c2-11e9-90f7-2975892eb186.png)

Hopefully, this gets resolved for the students wanting to take OSSU.

**Comment by ClockCat** [comment]
The content for the intro to programming course is only shown for the first lesson...but audit ends access next month and says you won't be able to view it any longer. I'm guessing next week the second lesson will show up as available....so it won't be possible to complete the course at all without paying for it?

**Comment by pachecohjr** [comment]
@ClockCat You could always make a dummy account.

**Comment by pachecohjr** [comment]
@ClockCat No, you should be able to view any of the weeks of lectures at any time. You work at your pace. Besides the grievances made here about the free audit access, once the two weeks are up, your "View Course" will be gone and you will no longer have access anymore.

To prevent that while you're in the middle of access / lectures, you can always make a dummy account...

**Comment by ahmedbenammar029** [comment]
you should add more Infos about the courses like its frequency like once a month twice ...
 also the access period especially with EDX's new audit policy

**Comment by waciumawanjohi** [comment]
@ahmedbenammar029 I agree that both of these ideas are useful information for students!

Re: _the access period especially with EDX's new audit policy_
The access period is determined for individual courses, edX has no site wide policy for how long access will last.
One possibility is to add a link to [this page](https://support.edx.org/hc/en-us/articles/360013426233-How-long-can-I-access-a-course-in-the-audit-track-) on the homepage somewhere. At the moment I don't have a clear idea of where that might fit. The homepage at the moment is focused on the high level concerns (which courses, what prereqs, etc) rather than low level details of individual course platforms. If you have ideas for where you think this link could naturally live on the homepage, definitely share!

Re: _add more Infos about the courses like its frequency like once a month twice_
On this, I suspect this information is unknowable. I haven't seen a schedule where any platform promises to hold new sessions on any particular schedule. The platforms just report when the next session is going to be held. And even for individual courses, the frequency can vary widely. I've seen a course (that we eventually de-listed for not being offered frequently enough) whose frequency varied between a few times a year to once every few years. Without clear communication from the platforms, any prediction of frequency on our part would just be a guess.

**Comment by ClockCat** [comment]
> @ClockCat No, you should be able to view any of the weeks of lectures at any time. You work at your pace. Besides the grievances made here about the free audit access, once the two weeks are up, your "View Course" will be gone and you will no longer have access anymore.
> 
> To prevent that while you're in the middle of access / lectures, you can always make a dummy account...

I can only view the first two weeks. The entire course is not available?

![image](https://user-images.githubusercontent.com/10076320/64566725-97c40480-d324-11e9-9c12-46a5feb0400b.png)

**Comment by ClockCat** [comment]
@pachecohjr

I can only view the first two weeks still.

![image](https://user-images.githubusercontent.com/10076320/64566907-0a34e480-d325-11e9-81a3-913a0c7b9d49.png)

**Comment by cipherblock** [comment]
@ClockCat I was pretty confused about this as well because I can also only see the first two weeks. When the course first began (on the 28th of August I believe), only Week 1 was there so I thought maybe the next week gets added every Wednesday. Is this not normally the case?

**Comment by waciumawanjohi** [comment]
@ClockCat @cipherblock
Thank you for taking the time to let us know about these changes. Which course is giving you access 1 week at a time? Is this Computing in Python from Georgia Tech? Intro to Computer Science from MIT? Or another course?

**Comment by cipherblock** [comment]
> @ClockCat @cipherblock
> Thank you for taking the time to let us know about these changes. Which course is giving you access 1 week at a time? Is this Computing in Python from Georgia Tech? Intro to Computer Science from MIT? Or another course?

Absolutely no problem. It's Intro to Computer Science and Programming using Python from MIT. I'm not 100% sure if that is what's happening because I'm not sure how it used to work. Maybe if I were to complete all available lectures before the next week began it would show up with the next week but I doubt it. I hope somebody else can shed more light on this.

**Comment by bradleygrant** [comment]
When I've taken MIT 600.1x in the past, they would release one week of material, one week at a time. That's true whether you audit or get verified for the class. The course calendar tells you what day they're going to drop each week's material.

Now, if they have 13 weeks worth of material to drop, and you only have 5 weeks of time to audit (I'm presuming this, I don't know), that it seems like you only get to see the first 5 weeks. So the thing to do would be enter the class in the last five weeks and try to do the whole thing in 5 weeks.

They weren't doing any of this a year ago, and I still have perpetual access to all of these introductory classes.

Would it be helpful at all if I were to try to life-raft some of this information into a free repository somewhere? I'm not sure the extent to which that can be done, as I'm pretty sure it's a licensed work even if it's not chargeable.

**Comment by bradleygrant** [comment]
One class that's still fully free to audit for the entire length of the course is UT Austin's "Linear Algebra: Foundations to Frontiers" on EdX. AND you get access to MATLAB for the duration of the course. Even if you audit for free.

**Comment by ivicts** [comment]
I have taken this course earlier this year (after Edx policy changed). If let's say the course is 13 weeks, you will have access from the start until the final exam (after 13th week). So, you do not get only 5 weeks. What you get depends on the course length.

**Comment by cipherblock** [comment]
Yes @bradleygrant I think @ivicts is correct, I've just checked 'Important Course Dates' and it's telling me that the course will end on 30th October.

On an unrelated note, could somebody help me to understand what the difference is between the 6.00.1x course on edx and the version that is on OCW because while doing the course on edx one of the community TA's advised somebody to watch this: https://ocw.mit.edu/courses/electrical-engineering-and-computer-science/6-0001-introduction-to-computer-science-and-programming-in-python-fall-2016/lecture-videos/lecture-1-what-is-computation/ 
So I'm wondering if they're the same/similar why is it being advised to also watch these lectures?

**Comment by hanjiexi** [comment]
I recently came across Princeton University's introductory CS class on Coursera:
- https://www.coursera.org/learn/cs-programming-java
- https://www.coursera.org/learn/cs-algorithms-theory-machines

Significantly, all features are available completely for free. Has anyone tried these courses?

**Comment by waciumawanjohi** [comment]
### Close of the Comment Period

**Findings**:
1) The feedback above highlighted the uncertainty that students feel around edX courses. Several students were unclear whether they would be able to view all of the lectures in the courses they were auditing. Several students expressed disappointment that they could not view the assessments for their edX course.

2) New courses were identified that may fit the goal of the Intro to Programming section. Questions about the courses quality remain unanswered.

3) Several students endorsed the course CS50. 

**Response**:
1) As students begin their educational journey to the equivalent of a bachelor's in Computer Science, OSSU should reduce student uncertainty. We should prioritize courses that offer students full access to materials. For that reason, _Computing in Python I: Fundamentals and Procedural Programming_ will not be added to Intro to Programming. If there is a change to edX's policies in the future, it would be wise to reevaluate this decision.

2) It is only possible to evaluate courses with feedback from participants. I look forward to reading such feedback for Princeton's new courses!

3) CS50 remains a popular course. It is also a course that begins by diving to a very low level of abstraction, teaching manual memory management by the 4th week. This is an increasingly rare approach, with CS departments increasingly using higher level languages in introductory courses. I have written in depth why I do not believe CS50 should be the first course in OSSU [here](https://github.com/ossu/computer-science/issues/440#issuecomment-397150554).

**Conclusion**
Intro to Programming will recommend two of the three proposed courses:

_Programming for Everybody
An Introduction to Interactive Programming in Python_
